core/templates/dev/head/domain/classifier/SVMPredictionService.js

```diff
@@ -0,0 +1,115 @@
+ * after training of classifier on Oppia-ml. If there are any changes in
+ * format of extracted data then corresponding changes must be propagated here.
+ * Oppia-ml uses scikit's SVC class during training classifier which uses
+ * libsvm's implementation. If there are any changes in following part of
+ * code in libsvm then corresponding changes must be propagated here.
+ * 
+ * libsvm's code for prediction:
+ * https://github.com/arnaudsj/libsvm/blob/master/svm.cpp#L2481
+ */
+
+oppia.factory('SVMPredictionService', [function() {
+  return {
+    kernel: function(kernelParams, supportVectors, input) {
+      var kernel = kernelParams.kernel;
+      var kvalues = [];
+
+      if (kernel == 'rbf') {
+        var gamma = kernelParams.gamma;
+        var vectorLength = input.length;
+        for(var i = 0; i < supportVectors.length; i++) {
+          var sum = 0;
+          for(var j = 0; j < input.length; j++) {
+            sum += Math.pow((supportVectors[i][j] - input[j]), 2);
+          }
+          kvalues.push(Math.exp(-gamma * sum));
+        }
+        return kvalues;
+      }
+    },
+
+    predict: function(classifierData, input) {
+      var nSupport = classifierData.n_support;
+      var supportVectors = classifierData.support_vectors;
+      var dualCoef = classifierData.dual_coef;
+      var intercept = classifierData.intercept;
+      var classes = classifierData.classes;
+      var kernelParams = classifierData.kernel_params;
+
+      var startIndices = [];
+      startIndices[0] = 0;
+      for(var i = 1; i < nSupport.length; i++) {
+        startIndices[i] = startIndices[i - 1] + nSupport[i - 1];
+      }
+
+      // Find kernel values for supportVectors and given input. Assumes that
+      // input has same dimension and data type as any of the supportVectors.
+      var kvalues = this.kernel(kernelParams, supportVectors, input);
+
+      var votes = [];
+      for(var i = 0; i < classes.length; i++) {
+        votes.push(0);
+      }
+
+      var p = 0;
+      for(var i = 0; i < classes.length; i++) {
+        for(var j = i + 1; j < classes.length; j++) {
+          var si = startIndices[i];
+          var sj = startIndices[j];
+          var ci = nSupport[i];
+          var cj = nSupport[j];
+
+          var coef1 = dualCoef[j - 1];
+          var coef2 = dualCoef[i];
+
+          var sum = 0;
+          for(var k = 0; k < ci; k++) {
+            sum += kvalues[si + k] * coef1[si + k];
+          }
+
+          for(var k = 0; k < cj; k++) {
+            sum += kvalues[sj + k] * coef2[sj + k];
+          }
+
+          sum -= intercept[p];
+          if (sum > 0) {
+            votes[i]++;
+          }
+          else {
+            votes[j]++;
+          }
+          p++;
+        }
+      }
+
+      // Find out class which has got maximum votes.
+      var maxVoteIdx = 0;
+      for(var i = 0; i < votes.length; i++) {
+        if (votes[i] > votes[maxVoteIdx]) {
+          maxVoteIdx = votes[i];
```

Comment:
Great! Btw, were any of the SVMPredictionService tests supposed to have caught this error?